src/highlights/highlight_grid.ts

```diff
@@ -1,345 +1,228 @@
+    // Flag to indicate if the grid needs to be redrawn
+    private isDirty = false;
+    // line number -> (hlId -> decoration options)
+    // Cache the decorations for each line to avoid recalculating them
+    private lineDecorationsCache: Map<number, Map<number, DecorationOptions[]>> = new Map();
+
+    constructor(
+        // Used to get the editor and viewport
+        private readonly gridId: number,
+        // Normalizes highlight IDs and provides decorators
+        private readonly groupStore: HighlightGroupStore,
+        // Used to get the editor from the grid ID
+        private readonly bufferManager: BufferManager,
+        // Used to get the viewport from the grid ID
+        private readonly viewportManager: ViewportManager,
+    ) {}
+
+    private get editor() {
+        return this.bufferManager.getEditorFromGridId(this.gridId)!;
     }
 
-    /**
-     * Checks whether or not the contents of a text cell should actually span two cells. For instance:
-     * a  length:1 width:1
-     * 你 length:1 width:2
-     * 🚀 length:2 width:2
-     * 🕵️ length:3 width:2
-     * ❤️ length:2 width:2
-     *
-     * @param cellText The cell text to check
-     * @returns Whether or not this cell is a double-width cell
-     */
-    static isDouble(cellText: string): boolean {
-        return wcswidth(cellText) === 2 || cellText.length > 1;
+    private get viewport() {
+        return this.viewportManager.getViewport(this.gridId);
     }
 
-    /**
-     * Get the width of a piece of text, in cells. Note that the use of tabs is
-     * position dependent, and this function must assume a line with tabs starts
-     * at position 0in the editor.
-     *
-     * @param text The text to get the width of
-     * @param tabSize The size of tab-widths
-     * @returns The width of the text, in cells
-     */
-    static getWidth(text: string, tabSize: number): number {
-        const t = expandTabs(text, tabSize);
-        return HighlightGrid.splitGraphemes(t).reduce((p, c) => p + (HighlightGrid.isDouble(c) ? 2 : 1), 0);
-    }
+    // #region Handle Redraw Events
 
-    static splitGraphemes(str: string): string[] {
-        return HighlightGrid.splitter.splitGraphemes(str);
+    handleGridLine(line: number, vimCol: number, cells: VimCell[]) {
+        // normalizes the highlight IDs
+        const vimCells = cells.map((cell) => {
+            const hlId = cell[1];
+            if (hlId) {
+                cell[1] = this.groupStore.normalizeHighlightId(hlId);
+            }
+            return cell;
+        });
+        this.gridLine.handleGridLine(line, vimCol, vimCells);
+        this.lineDecorationsCache.delete(line);
+        this.isDirty = true;
     }
 
-    cleanRow(row: number) {
-        delete this.grid[row];
+    handleRedrawFlush() {
+        if (this.isDirty && this.editor && this.viewport) {
+            this.refreshDecorations();
+            this.isDirty = false;
+        }
```

Comment:
It might be worth an info log if there is no editor or viewport; it may be helpful in debugging user issues

Reply:
Regarding this, I have added detailed comments. It is expected for the editor to be null here. In fact, every time a new window is created, the editor is null.